[RFC 0080] Change NixOS releases to YY.05,YY.11

jonringer opened the pull request:
This is meant to help minimize risk and work during the release stabilization period, by selecting a better time to branch-off from unstable.

Adapted from https://github.com/NixOS/rfcs/commit/8a3ba49393455fcf426488aef8b6d27d428c640c

Related discussions: 
 - https://discourse.nixos.org/t/20-09-release-retrospective/9851
 - https://discourse.nixos.org/t/what-should-stable-nixos-prioritize/9646

cc @worldofpeace @garbas @FRidh 

rendered: https://github.com/NixOS/rfcs/blob/ca06bb8dbb245f6a4c00311967e37753cd3d691a/rfcs/0080-nixos-release-schedule.md

Ma27 commented:
Another benefit of that would be that the stabilization period wouldn't clash so much with the exam phases (at my university at least).

jonringer commented:
I will write up another one for proposing a change to the ZHF schedule, but those changes are more controversial and didn't want to combine the two, as it would increase the "surface area" of changes greatly.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/release-schedule-for-nixos/3149/11

ryantm commented:
I nominate myself for shepherd and/or shepherd leader.

domenkozar commented:
Happy to shepherd this one to get it speed it up.

ryantm commented:
Everyone, please self nominate as shepherd. We want to get this through as quickly as is reasonable so we can minimize uncertainty.

ryantm commented on rfcs/0080-nixos-release-schedule:
```diff
@@ -0,0 +1,69 @@
+# Motivation
+[motivation]: #motivation
+
+By changing our release dates to follow those of certain upstream projects
```
```suggestion
By changing our release dates to follow those of certain upstream projects,
```

ryantm commented on rfcs/0080-nixos-release-schedule:
```diff
@@ -0,0 +1,69 @@
+
+By changing our release dates to follow those of certain upstream projects
+we could reduce the amount of issues we encounter during stabilization.
+In particular, Gnome and KDE Plasma both have a release in October, which
```
```suggestion
In particular, GNOME and KDE Plasma both have a release in October, which
```

ryantm commented on rfcs/0080-nixos-release-schedule:
```diff
@@ -0,0 +1,69 @@
+we could reduce the amount of issues we encounter during stabilization.
+In particular, Gnome and KDE Plasma both have a release in October, which
+allows ample time for NixOS contributors to package and stabilize the
+desktop managers before the November release. For the May release, Gnome will
```
```suggestion
desktop managers before the November release. For the May release, GNOME will
```

ryantm commented on rfcs/0080-nixos-release-schedule:
```diff
@@ -0,0 +1,69 @@
+they will be compatible with the version of systemd during branch-off.
+
+The current March and September release months may be the worst months for
+Gnome and Plasma, as the release available during stabilization would be
```
```suggestion
GNOME and Plasma, as the release available during stabilization would be
```

ryantm commented on rfcs/0080-nixos-release-schedule:
```diff
@@ -0,0 +1,69 @@
+however, plasma-5.19 and the yet-released 5.20 both supported systemd-246 with
+zero additional effort.
+
+By choosing branch-off points in master where the major desktop manager use
```
```suggestion
Choosing branch-off points in master where the major desktop manager use
```

ryantm commented on rfcs/0080-nixos-release-schedule:
```diff
@@ -0,0 +1,69 @@
+## Core changes
+
+The 21.03 release will be delayed until 21.05, and the current 20.09 release
+will be supported until June 2021. After the initial 21.05 release, all subsequent
```
Specifying the exact time 20.09 is supported until is a change of policy. The current policy is to keep the previous release supported for 1 month after the next release.

jonringer replied:
This is in regards to https://nixos.org/manual/nixos/unstable/release-notes.html#sec-release-20.09 in which we say April of 2021

ryantm reviewed (changes requested):
Please add .md as the file extension.

jtojnar commented on rfcs/0080-nixos-release-schedule:
```diff
@@ -0,0 +1,69 @@
+
+By changing our release dates to follow those of certain upstream projects,
+we could reduce the amount of issues we encounter during stabilization.
+In particular, GNOME and KDE Plasma both have a release in October, which
```
GNOME 3.38 released on [September 16](https://wiki.gnome.org/Forty), we only merged it on [October 24](https://github.com/NixOS/nixpkgs/pull/98316#event-3916100085) and that included only rough testing by the GNOME packagers team (all core apps start and we did not experience any serious issues when running GNOME session). It was merged to master on [November 10](https://github.com/NixOS/nixpkgs/pull/102739#event-3977835342), which allowed wider user base to start testing this. That revealed a [serious regression](https://github.com/NixOS/nixpkgs/issues/103746) that GNOME maintainers were not able to reproduce as of yet. With this proposal a branch-off would be somewhere around the end of October so I doubt the delay will be enough to produce a well tested and stable release.

jonringer replied:
historically, the branch-off was done in the middle of the month, so 20.03 I think was around the ~14th, for 20.09 it was the beginning of the second second week, so I think around the 7th. And the target release date is usually on the last week of the month.

I have another RFC proposal (which I am in the process of creating) which would create a schedule for doing the desktop manager tasks before ZHF. Remember that the majority of the work for GNOME 3.38 was done while ZHF was still going on, and efforts were divided. If we have a few weeks were the community is focused on packaging plasma and gnome, I don't think the time to package them will be so long.

Also, organizing the effort with materials for people to "onboard" onto the effort helps a lot, something like https://github.com/NixOS/nixpkgs/issues/101964 where I wrote up  how to go about doing the fixes, and then tracked the overall progress of the effort.

jonringer replied:
I fixed the release month from October to September (not sure why I put October).

Atemu replied:
@jtojnar Would it perhaps be possible to start the GNOME packages update a few weeks before release based on the (at that time) current development versions?

Would that make the bump of the actual release small enough to allow for a tighter window?

jtojnar replied:
Last few cycles we actually started packaging beta or alpha releases. The updates themselves are mostly automated so the biggest time sink is going through all the changelogs. We eventually want to have a nixpkgs branch tracking git `HEAD`s of all the GNOME packages but need a tooling first – with unreleased code, we do not have the benefit of curated changelogs.

garbas commented:
Happy to help with this one. I'd like to nominate myself as a shepherd.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/20-09-release-retrospective/9851/30

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/what-should-stable-nixos-prioritize/9646/77

jtojnar commented on rfcs/0080-nixos-release-schedule:
```diff
@@ -0,0 +1,69 @@
+
+# Future work
+[future]: #future-work
+
```
- Update [osinfo-db entry](https://gitlab.com/libosinfo/osinfo-db/-/blob/master/data/os/nixos.org/nixos-unstable.xml.in) with the changes

jonringer replied:
Already captured in https://github.com/NixOS/nixpkgs/issues/97405 and outside of the scope of the RFC

jtojnar replied:
The issue only seems to deal with release process. I assume this RFC will mean changing https://github.com/NixOS/nixpkgs/blob/master/.version, which means we will need to update the osinfo-db entry for `nixos-unstable` to match.

jonringer replied:
yes, this is mentioned here: https://nixos.org/manual/nixos/stable/index.html#at-beta-release-time with the heading "On the master branch, increment the .version file".

I will update the example command when transferring over to the https://github.com/NixOS/release-wiki

jtojnar replied:
Well, that is still only relevant to the 21.11 bump. Or will the unstable branch remain 21.03pre until then, even after this RFC has been accepted?

jonringer replied:
>  Or will the unstable branch remain 21.03pre until then, even after this RFC has been accepted?

unstable branch will always be the pre suffix

only release branches will differ

jtojnar replied:
Yes. But it is not clear to me whether after accepting this RFC, the unstable branch will become `21.05pre` instead of `21.03pre` it currently is.

This is orthogonal to the beta release and will affect whether we have to update the osinfo-db.

jonringer replied:
Yes, 21.03pre will need to be changed to 21.05pre. Sorry.

jonringer replied:
added the related two work items. Sorry for misinterpreting.

edolstra commented:
This RFC is now in discussion with @ryantm, @domenkozar and @garbas as shepherds. @ryantm Do you want to be shepherd leader on this one?

ryantm commented:
Thank you, @edolstra. I accept being the shepherd leader.

It looks like the one concern raised so far is by @jtojnar that the delay will not be enough to get GNOME ready (which is a major reason for this RFC). It seems like @jonringer has answered that he still intends branch off to be around the time when GNOME was merged into master on the last cycle.

In the interest of limiting uncertainty in the community about release dates, I will lead this shepherd team as quickly as is allowed by the RFC rules (but not any quicker).

I motion for the Final Comment Period (FCP) to begin with disposition to merge. @domenkozar @garbas please either approve or disapprove of the motion as soon as you can. I understand this is a bit fast, and I won't mind if you disapprove.

jonringer commented:
Squashed fixup commits, and added `.md` file extensions so that the document could be rendered: https://github.com/NixOS/rfcs/blob/ca06bb8dbb245f6a4c00311967e37753cd3d691a/rfcs/0080-nixos-release-schedule.md

domenkozar commented:
I approve for FCP. Anything that makes releasing easier is a good step forward.

We don't know the unknowns that come up with releases in May/November, but then we can make another shift if needed (hopefully not).

ryantm commented:
@garbas Could you please approve or disapprove of my motion to move to FCP with disposition to merge?

jonringer commented:
> We don't know the unknowns that come up with releases in May/November, but then we can make another shift if needed (hopefully not).

I think unstable + ZHF will do enough to stabilize releases for most production users regardless of timeline. The main issue is trying to provide a decent desktop experience. And seeing as how it is usually the biggest unknown going into a release, we should try to minimize it as much as possible; which is the goal of this RFC. At the very least, it should be better than current release schedule.

One other thing we could do is also define release criteria, and that should go a long way to determining what we prioritize, but I don't think anyone will disagree that we should have usable graphical installers and related desktop managers.

garbas commented:
I'm also in favor to move this to FCP. I don't see any drawbacks, except minor inconveniences.

ryantm commented:
We are now in the Final Comment Period (FCP). Unless any major objections are raised, this will be accepted and merged in 10 days from now, on December 15th!

Please post any final comments on this pull request.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/rfc-0080-fcp-change-nixos-releases-to-yy-05-yy-11/10373/1

Lassulus commented:
Not sure if this is the right place to voice my opinion, but personally I would be happy with a longer release period. Like once a year maybe? I don't have any deep insights, but I feel like a longer release period would be less straining on release managers (could also be the other way around, since they would be bound for a longer time?). Also longer release cycles would be less work downstream on users and other people.

Anyway, I'm glad seeing so many people invested in this great operation system <3

jonringer commented:
> Not sure if this is the right place to voice my opinion, but personally I would be happy with a longer release period. Like once a year maybe?

I think @nh2 said [it best](https://discourse.nixos.org/t/what-should-stable-nixos-prioritize/9646/69).
> I think 6-months releases are the sweet spot across the effort to upgrade, easy security patches backporting, and having new enough software for businesses.

A lot of work goes on in unstable, and increasing the length of stable releases will probably increase the maintenance burden. Not to mention that ZHF helps unstable as well with reducing the number of failing builds and tests

rnhmjoj commented:
I think it's a good idea: it's not just DEs the follow this schedule. I maintain quite a bit of packages and several projects release big updates around march and september, often *after* the feature freeze. 
This forces me to either make two PR (against master and the release branch) or backport the updates, which doubles the work even if it's trivial in many cases.

ryantm commented:
The FCP has ended without additional objections or blocking comments. I declare this RFC accepted. @NixOS/rfc-steering-committee please merge this.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/what-should-stable-nixos-prioritize/9646/78

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/21-05-call-for-release-manager/10204/8

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/21-05-call-for-release-manager-and-release-members/10204/17

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/21-05-has-been-released/13407/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/rfc-0094-fcp-use-matrix-for-official-chat/14631/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/what-should-stable-nixos-prioritize/9646/81
